### PR TITLE
fix: Aggregate Serenity brand-presence endpoints across all models when platform is omitted | LLMO-7093

### PR DESCRIPTION
**Pull request by igor-grubic**
## Problem

The Serenity brand-presence Elements payload builders resolved the model via `resolveElementModel(model || platform)` and **always** added a `CBF_model` filter. `resolveElementModel` falls back to `search-gpt` for an absent or unrecognized value, so there was no way to express "all models" on the wire — the "All Platforms" view silently scoped every query to a single model instead of aggregating.

## Changes Made

- Added `isAllModelsFilter()` (true when the value is absent, empty, or the `all` sentinel) and a `buildModelFilter()` helper in `src/support/elements/constants.js`.
- Wired both into the brand-presence builders (`brand-presence-stats` ×4, `kpi-headlines`, source-visibility, `sentiment-overview`, `market-tracking-trends`/`competitor-summary`) so `CBF_model` is **omitted** for "All Platforms". Semrush then aggregates server-side across every model the brand has data for.
- Added `buildAdvancedFilters()`, which drops the `advanced` key entirely when the filter list ends up empty — Semrush returns HTTP 422 on an empty AND block (see *Live verification* below).
- Updated the OpenAPI model-param docs and added `all` to the canonical values for the six affected operations.

This is the same pattern the sibling `url-prompts` endpoint already ships, and what the pre-existing `isAllPlatforms`/`ALL_PLATFORMS` sentinel was staged for. Because Semrush computes the cross-model aggregate itself, there is no in-code weighting to get wrong — this **resolves the ticket's open question about aggregation math**.

The single-model path (`platform=<model>`) and every other endpoint that legitimately defaults to `search-gpt` (prompts, topics, cited-domains, owned-urls, url-inspector) are unchanged.

## Related Issues

- Jira: https://jira.corp.adobe.com/browse/LLMO-7093
- Slack thread: https://cq-dev.slack.com/archives/C0BEJH1SNNL/p1787056409321099
- **Frontend half: adobe/project-elmo-ui#2888 — still OPEN (approved + clean). See *Merge order* below.**

## ⚠️ Merge order

These two PRs must land **together**, and the UI must never land first:

| Deployed | UI sends | API resolves to | Brand Visibility (Asian Paints) |
|---|---|---|---|
| Neither (today) | `platform=openai` | `chatgpt-paid` | **46.3%** ❌ |
| API only | `platform=openai` | `chatgpt-paid` | 46.3% — no user-visible change |
| **UI only** | *(omitted)* | `search-gpt` (fallback) | **62.6%** ❌ silently switches to a *different* wrong model |
| **Both** | *(omitted)* | no `CBF_model` | **55.3%** ✅ matches the Semrush MFE |

## Live verification (done)

Verified against the production Semrush Elements API by driving the embedded Brand Presence MFE with Playwright and rewriting its request payloads (the browser supplies auth). Brand **Asian Paints**, sub-workspace `c8feffff-6e58-41db-b804-5652033dd292`, project IN-en `edf08d1d-00dd-463c-bb59-7fdd8445acf1`, `2026-08-04`→`2026-09-02`. All calls were read-only queries. Full evidence in [this comment](https://github.com/adobe/spacecat-api-service/pull/3094#issuecomment-5507572042).

**1. Omitting `CBF_model` is exactly equivalent to enumerating every model** — ratios *and* raw counts, across six elements:

| element | `search-gpt` | **omit `CBF_model`** | explicit OR-list of all 10 models |
|---|---|---|---|
| `TOTAL_EXECUTIONS` `601590e0` | 3,980 | **39,725** | 39,725 |
| `MENTIONS` `e1a6811b` | 5,607 | **48,638** | 48,638 |
| `CITATIONS_KPI` `588054fe` | 54,658 | **147,956** | 147,956 |
| `VISIBILITY` `2724878e` | 0.6263819095477386 | **0.5525739458779106** | 0.5525739458779106 |
| `KPI_BRAND_VISIBILITY` `6db33cf0` | 0.6263819095477386 | **0.5525739458779106** | 0.5525739458779106 |
| `KPI_SHARE_OF_VOICE` `69a4befb` | 0.5503322167381302 | **0.3135053764357032** | 0.3135053764357032 |

Identical to full float precision, and clearly *not* byte-identical to `platform=search-gpt`. The **count** metrics matching confirms Semrush returns a deduped cross-model union rather than summing per-model rows — no inflated executions/mentions/citations.

**2. Overview and Brand Presence already agree.** The two surfaces hit *different* elements (Overview's `/stats` uses `VISIBILITY` `2724878e`; Brand Presence uses `KPI_BRAND_VISIBILITY` `6db33cf0`, and the MFE never calls the former), yet they return bit-identical values for every model filter tested. So there is no per-surface aggregation drift — the `all` → single-model coercion was the *only* discrepancy. The customer-reported 55% Brand Visibility in LLMO-7058 reproduces exactly.

**3. Empty AND is NOT "match all" — the earlier assumption in this PR was wrong.** Semrush rejects it:

| element | `advanced: {op:'and', filters: []}` | `advanced` key omitted |
|---|---|---|
| `SENTIMENT` `f4153af8` | **422** `{"message":"request could not be processed"}` | 200 |
| `TRENDS_MV` `b5281393` | **422** | 200 |
| `MARKET_CITATIONS_TREND` `2e5a6f4e` | **422** | 200 |

Disambiguated: an `advanced` block carrying only `CBF_model` (no `CBF_project`) returns 200, so the 422 comes from the empty AND itself, not a missing required filter. This was **production-reachable** — the Overview-SR sentiment card requests all platforms with no region and no category, so the default sentiment view would have broken. Fixed in `2ba27cff5`; the exact patched builder output now returns 200 with real data (16 and 17 series rows) on all three elements.

**4. One deviation from the MFE (non-blocking).** The MFE does not omit the filter — it sends an explicit `or` block enumerating all 10 model names. Omission is exactly equivalent here (table 1), so this PR is correct as written. The two would only diverge for a brand holding historical data from a model no longer in its enabled set. Candidate follow-up if we want MFE parity by construction rather than by observation.

## Testing

- `npm test` — targeted element suites pass (96 assertions across `constants`, `brand-presence-stats`, `kpi-headlines`, `market-tracking-trends`, `sentiment-overview`), including new regression coverage for the empty-`advanced` guard.
- Live Elements API verification as above.

## Out-of-scope follow-up (open question)

`/brand-presence/topics` and `/brand-presence/topics/:topicId/prompts` also coerce an omitted platform to single-model `search-gpt` and do not handle the `all` sentinel, but they are not in LLMO-7093's explicit affected-endpoints list, so they were intentionally left out. Decide whether "All Platforms" should also cover the Data Insights (topics) surface; if yes, a follow-up would apply the same `buildModelFilter` wiring after confirming per-topic aggregation semantics.

## Change Management

```yaml
cm-assessment: v1
changeType: standard
impact: unnoticeable
risk: minor
changeApprovedBy: ["Dominique Jäggi", "Vivek Singh"]
rationale: "Corrects an aggregation bug in internal API response values (extensively live-verified); coordinated-deploy note concerns an internal UI consumer, not an external/public API contract."
```

### CODE REVIEW

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @igor-grubic,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Request changes** - one unverified edge case needs acknowledgement before merge.
**Complexity: HIGH** - medium diff; API surface.
_Changes: Fixes brand-presence endpoints to aggregate across all AI models when the platform parameter is omitted, instead of silently defaulting to search-gpt (11 files)._
Note: Recommend a human read before merge - this change modifies a shared API contract (OpenAPI spec). The bot review is a complement to, not a replacement for, a human read here.

## Must fix before merge
1. [Important] Empty `advanced:{op:'and',filters:[]}` edge case extends beyond sentiment-overview to market-tracking-trends/competitor-summary - `src/support/elements/definitions/market-tracking-trends.js:73` (details inline)

<details>
<summary>Non-blocking (3): minor issues and suggestions</summary>

- suggestion: Add direct unit tests for `buildModelFilter` in `test/support/elements/constants.test.js` - the function has non-trivial branching (null path, wrap option, model resolution) and is only tested indirectly through builder tests. Codecov confirms coverage, but direct tests pin the contract independently of downstream consumers.
- nit: Existing test titles "always includes CBF_ws_brand and CBF_model" in `test/support/elements/definitions/brand-presence-stats.test.js` are now misleading since CBF_model is conditional. Consider "includes CBF_model when a concrete model is provided".
- nit: `isAllModelsFilter` accepts non-string falsy values (0, false) as "aggregate" via the `typeof value !== 'string'` check. Not a practical bug (HTTP params are always strings or undefined), but a test case documenting this behavior would make the contract explicit.
</details>

File: src/support/elements/definitions/market-tracking-trends.js [hunk 1]
```diff
@@ -47,7 +47,9 @@ import { dateToIsoWeek } from '../week-utils.js';
  * `comparison_data_formatting` (the MFE omits it for these elements).
  *
  * @param {object} params
```
Comment (MysticatBot) [inline src/support/elements/definitions/market-tracking-trends.js:49]: issue (blocking): The PR description flags the empty advanced filter edge case for live verification on sentiment-overview only. However, buildMarketTrendPayload (which backs both market-tracking-trends and competitor-summary) has the same exposure: when model is absent AND projectIds is empty, advancedFilters resolves to [].

This matters because if Semrush interprets an empty and block as match nothing rather than no constraint, the aggregate view would silently return zero data.

**Fix:** Either (a) extend your live verification plan to cover market-tracking-trends and competitor-summary with the all-platforms + no-project-filter combination, or (b) harden buildMarketTrendPayload to omit the advanced key entirely when advancedFilters.length === 0 (a one-line guard that protects both endpoints and any future builder that follows this pattern).

Comment (igor-grubic) [reply src/support/elements/definitions/market-tracking-trends.js:51]: Confirmed — you were right, and it is worse than "an edge case". Fixed in 2ba27cff5.

I verified this live against the Semrush Elements API (brand **Asian Paints**, sub-workspace `c8feffff-6e58-41db-b804-5652033dd292`, `2026-08-04`→`2026-09-02`), by intercepting the embedded MFE's own requests and rewriting the payloads so the browser supplied auth.

**An empty AND is NOT "match all" — Semrush rejects it:**

| element | `advanced: {op:'and', filters: []}` | `advanced` key omitted |
|---|---|---|
| `SENTIMENT` `f4153af8` | **422** `{"message":"request could not be processed"}` | 200 |
| `TRENDS_MV` `b5281393` | **422** | 200 |
| `MARKET_CITATIONS_TREND` `2e5a6f4e` | **422** | 200 |

I also disambiguated the cause: an `advanced` block carrying **only** `CBF_model` (no `CBF_project`) returns **200**, so the 422 comes from the empty AND itself, not from a missing required project filter. Dropping the key is therefore the correct and sufficient fix.

**This was production-reachable, not theoretical.** The live Overview-SR sentiment card requests `sentiment-overview?platform=openai&startDate=…&endDate=…` — no `projectId`, no `category`. With this PR + adobe/project-elmo-ui#2888, that becomes no-model / no-project / no-category → empty AND → 422, breaking the default sentiment view.

The fix centralises the guard in `buildAdvancedFilters` (next to `buildModelFilter`) and applies it to both affected builders. I audited the rest of the family: only `market-tracking-trends` and `sentiment-overview` can reach an empty list — the stats and kpi-headline builders always seed a mandatory brand filter.

Final check: the exact patched builder output returns **200 with real data** (16 and 17 series rows) on all three elements, while the unpatched output returns 422.

**Comment by igor-grubic** [comment]
## Live verification against the Semrush Elements API

The "omit `CBF_model` → Semrush aggregates across all models" premise was previously an untested assumption. I've now verified it empirically, so it's no longer a guess.

**Method:** drove the production Brand Presence page (`abv.adobe.com`, Asian Paints) with Playwright, intercepted the embedded Semrush MFE's own Elements requests, and rewrote the payloads — the browser supplies auth, so no separate credential is needed. All calls are read-only queries. Brand **Asian Paints**, sub-workspace `c8feffff-6e58-41db-b804-5652033dd292`, project IN-en `edf08d1d-00dd-463c-bb59-7fdd8445acf1`, `2026-08-04`→`2026-09-02`.

### 1. Omitting `CBF_model` is exactly equivalent to enumerating every model ✅

Tested across six elements, ratios *and* raw counts:

| element | `search-gpt` | **omit `CBF_model`** | explicit OR-list of all 10 models |
|---|---|---|---|
| `TOTAL_EXECUTIONS` `601590e0` | 3,980 | **39,725** | 39,725 |
| `MENTIONS` `e1a6811b` | 5,607 | **48,638** | 48,638 |
| `CITATIONS_KPI` `588054fe` | 54,658 | **147,956** | 147,956 |
| `VISIBILITY` `2724878e` | 0.6263819095477386 | **0.5525739458779106** | 0.5525739458779106 |
| `KPI_BRAND_VISIBILITY` `6db33cf0` | 0.6263819095477386 | **0.5525739458779106** | 0.5525739458779106 |
| `KPI_SHARE_OF_VOICE` `69a4befb` | 0.5503322167381302 | **0.3135053764357032** | 0.3135053764357032 |

Identical to full float precision in every case. Notably the **count** metrics match too — Semrush returns a deduped cross-model union, it does **not** sum rows per model. That closes the main risk with this approach (no inflated executions/mentions/citations).

### 2. Overview and Brand Presence already agree — the only discrepancy is the `all` → `openai` coercion ✅

Worth calling out because the two surfaces hit *different* Semrush elements: Overview's `/brand-presence/stats` uses `VISIBILITY` `2724878e`, while Brand Presence uses `KPI_BRAND_VISIBILITY` `6db33cf0`, and the MFE never calls the former. They nevertheless return bit-identical values:

| model filter | Overview `VISIBILITY` | Brand Presence `KPI_BRAND_VISIBILITY` |
|---|---|---|
| `chatgpt-paid` | 0.46256281407035177 | 0.46256281407035177 |
| `search-gpt` | 0.6263819095477386 | 0.6263819095477386 |
| omitted | 0.5525739458779106 | 0.5525739458779106 |

So there is no per-surface aggregation drift to reconcile. Concretely, for Asian Paints:

- **today** — Overview "All Platforms" sends `platform=openai`, which `resolveElementModel` maps to `chatgpt-paid` → **46.3%**
- **Semrush MFE**, all models → **55.3%**
- **with this PR + adobe/project-elmo-ui#2888** → **55.3%** ✅

That 9-point gap closing is exactly the LLMO-7058 complaint. (The customer-reported 55% Brand Visibility reproduces precisely.)

### 3. One correction to the approach

The MFE does **not** omit the filter — with all models selected it sends an explicit one-level `or` block enumerating all 10 model names. Omission happens to be exactly equivalent here (table 1), so this PR is correct as written. The two would only diverge for a brand holding historical data from a model no longer in its enabled set. Worth a follow-up if we want MFE parity by construction rather than by observation; not a blocker.

### 4. Empty AND → 422 (the blocking review comment)

Confirmed and fixed in 2ba27cff5 — details in the [inline thread](https://github.com/adobe/spacecat-api-service/pull/3094#discussion_r3912754988). Summary: `advanced: {op:'and', filters: []}` returns HTTP 422 on `SENTIMENT`, `TRENDS_MV` and `MARKET_CITATIONS_TREND`; omitting the key returns 200 on all three. This was reachable by the default Overview-SR sentiment view.

---

**Note:** the PR description says adobe/project-elmo-ui#2888 is "already merged" — it is still open (approved and clean). These two need to land together: this PR alone changes nothing user-visible, since the UI still coerces `all` → `openai` before the request is made.

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @igor-grubic,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Request changes** - one test-quality gap to close before merge.
**Complexity: HIGH** - large diff; API surface.
_Changes: Fixes brand-presence endpoints to aggregate across all AI models when the platform parameter is omitted or set to `all`, instead of silently defaulting to search-gpt (11 files)._
Note: Recommend a human read before merge - this change modifies a shared API contract (OpenAPI spec). The bot review is a complement to, not a replacement for, a human read here.

## Must fix before merge
1. [Important] `buildModelFilter` is exported and used by all five builders but has no direct unit tests - all coverage is indirect - `test/support/elements/constants.test.js` (details inline)

<details>
<summary>Non-blocking (4): minor issues and suggestions</summary>

- suggestion: OpenAPI parameter descriptions reference internal implementation details (`CBF_model`, Semrush server-side aggregation) - consumer-facing docs should describe behavior, not internals - `docs/openapi/serenity-api.yaml`
- suggestion: The two "absent platform" semantics (`isAllModelsFilter` for brand-presence family vs `resolveElementModel` for url-prompts/topics) are enforced only by which helper each definition file calls - a brief decision table in `constants.js` JSDoc or `definitions/README.md` would prevent a future contributor from picking the wrong one
- nit: `isAllModelsFilter` accepts non-string falsy values (0, false) as "aggregate" via the `typeof` check - not reachable from HTTP params but a test case documenting this would make the contract explicit - `src/support/elements/constants.js:136`
- nit: The `for` loop over both builders in "omits the advanced block entirely when all-platforms is combined with no region" means a failure in the second builder is harder to diagnose from the test name alone - consider splitting into two named cases - `test/support/elements/definitions/market-tracking-trends.test.js`
</details>

## Previously flagged, now resolved
- Empty `advanced` block with empty AND filters edge case confirmed worse than anticipated (Semrush returns HTTP 422, not "match all") - fixed in 2ba27cff5 with centralized `buildAdvancedFilters` guard and live Semrush API verification across SENTIMENT, TRENDS_MV and MARKET_CITATIONS_TREND elements.

File: test/support/elements/constants.test.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import { expect } from 'chai';
+import {
+  ALL_PLATFORMS,
+  isAllPlatforms,
+  isAllModelsFilter,
+  buildAdvancedFilters,
```
Comment (MysticatBot) [inline test/support/elements/constants.test.js:19]: issue (blocking): `buildModelFilter` is the central new public export in this PR - it has three distinct return paths (null for aggregate, or-wrapped eq for `wrap: true`, bare eq for `wrap: false`) plus model resolution via `resolveElementModel`. All paths are exercised indirectly through the builder tests (and codecov confirms line coverage), but there is no direct `describe("buildModelFilter", ...)` block.

Direct tests improve diagnosis: if a future change breaks the `wrap: false` path, the failing test currently names a builder, not this function. Adding ~15 lines covering null return, wrapped shape, bare shape, and platform-code translation would document the contract independently of its callers.

**Review by vivesing — APPROVED** [review]

File: test/support/elements/constants.test.js (hunk 2 above)
Comment (igor-grubic) [reply test/support/elements/constants.test.js:19]: Fair point — all coverage was indirect. Added a direct `describe('buildModelFilter')` in `9c2067f16` covering the four distinct paths:

- `null` return for the aggregate case (absent, empty, `'all'`, `ALL_PLATFORMS`) — including that it wins regardless of `wrap`
- the default wrapped one-member `or` block
- the bare `eq` under `wrap: false` (stats mentions/citations shape)
- UI platform-code translation (`openai` → `chatgpt-paid`, `chatgpt` → `search-gpt`, `gemini` → `gemini-2.5-flash`)
- unrecognized value → falls back to the default model, **not** the aggregate (the distinction that matters most here)

**Comment by igor-grubic** [comment]
All five items from the review are addressed in `9c2067f16`.

**Must fix**
- ✅ **`buildModelFilter` had no direct tests** — added a dedicated `describe` block covering all four return paths plus the unrecognized-value fallback. Details in the [inline thread](https://github.com/adobe/spacecat-api-service/pull/3094#discussion_r3913968375).

**Non-blocking — all four taken**
- ✅ **OpenAPI leaked internals.** Agreed, `CBF_model` is an upstream column name with no meaning to a consumer. Replaced across all six parameter descriptions with the behavioural contract instead: *"computed upstream as a deduplicated cross-model union (not a client-side sum)"*. That phrasing is worth keeping because the dedup-vs-sum distinction is exactly what a consumer needs to know, and it's the property I verified live (omitting the filter returns 39,725 executions — identical to enumerating all 10 models, not 10× a single model).
- ✅ **Two coexisting "absent platform" semantics.** Good catch — this is a real footgun, since the only thing selecting between them is which helper a definition file happens to call. Documented in the `buildModelFilter` JSDoc: `buildModelFilter` (absent/`all` → omit → aggregate) for the brand-presence family whose UI has a real "All Platforms" option, vs `resolveElementModel` (absent → `search-gpt`) for url-prompts/topics/cited-domains/owned-urls/url-inspector, which have no such option. Kept it in `constants.js` next to both helpers rather than a separate README so it sits where the choice is actually made.
- ✅ **`isAllModelsFilter` and non-string falsy values.** Added a test documenting that `0`, `false` and `{}` are treated as "absent". Not reachable from HTTP params (always `string | undefined`), but now explicit for non-HTTP callers.
- ✅ **Two-builder loop in the empty-`advanced` test.** Split into two named cases so a failure identifies the element.

**On the degraded-review notice:** there is no spec document for this change — it originated as a bug fix from a customer report (LLMO-7058) rather than a design. What the PR lacked was empirical validation of its central assumption, and that has now been supplied: the "omit `CBF_model` → Semrush aggregates" premise and the empty-AND behaviour were both verified live against the production Elements API. Full evidence in the [PR description](https://github.com/adobe/spacecat-api-service/pull/3094#issue) and [this comment](https://github.com/adobe/spacecat-api-service/pull/3094#issuecomment-5507572042).

Also noting your recommendation for a human read on the OpenAPI contract change — @vivesing has since approved.

Verification: 103 targeted assertions pass, ESLint clean, `npm run docs:lint` valid with no new warnings (220, unchanged).

**Review by MysticatBot — APPROVED** [review]
Hey @igor-grubic,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Approve** - all prior findings addressed, no remaining blocking issues.
**Complexity: HIGH** - large diff; API surface.
_Changes: Fixes brand-presence endpoints to aggregate across all AI models when the platform parameter is omitted or set to `all`, instead of silently defaulting to search-gpt (11 files)._
Note: Recommend a human read before merge - this change modifies a shared API contract (OpenAPI spec). The bot review is a complement to, not a replacement for, a human read here.

<details>
<summary>Non-blocking (1): minor issues and suggestions</summary>

- nit: OpenAPI parameter descriptions use inconsistent punctuation across endpoints - four endpoints end the aggregation clause with a period and uppercase continuation (`client-side sum). An unrecognized value...`), while kpi-headlines and source-visibility-headline use a semicolon and lowercase (`client-side sum);\n an unrecognized value...`) - `docs/openapi/serenity-api.yaml`
</details>

## Previously flagged, now resolved
- `buildModelFilter` now has direct unit tests covering all four return paths (null/aggregate, wrapped or block, bare eq, platform-code translation) plus the unrecognized-value fallback.
- OpenAPI descriptions no longer leak internal implementation details (`CBF_model`); they now describe the behavioral contract.
- Two coexisting "absent platform" semantics (`buildModelFilter` vs `resolveElementModel`) documented in JSDoc.
- `isAllModelsFilter` non-string falsy value behavior now has explicit test coverage.
- Two-builder loop in market-tracking-trends test split into named cases for clearer failure diagnosis.

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 2m 27s | <b>Cost:</b> $3.87 | <b>Commit:</b> <code>9c2067f1661313c50d9f5e8d3f8f56a3a04c2b4b</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>
